controllers/servers/csi/csi_controller_server.py

```diff
@@ -620,6 +620,57 @@ def DeleteVolumeGroup(self, request, _):
 
         return csi_pb2.DeleteVolumeGroupResponse()
 
+    def _add_volumes_missing_from_group(self, array_mediator, volume_ids_in_request, volume_ids_in_volume_group,
+                                        volume_group_id):
+        for volume_id in volume_ids_in_request:
+            if volume_id not in volume_ids_in_volume_group:
+                array_mediator.add_volume_to_volume_group(volume_group_id, volume_id)
+
+    def _remove_volumes_missing_from_request(self, array_mediator, volume_ids_in_request, volume_ids_in_volume_group,
+                                             volume_group_id):
+        for volume_id in volume_ids_in_volume_group:
+            if volume_id not in volume_ids_in_request:
+                array_mediator.remove_volume_from_volume_group(volume_group_id, volume_id)
+
+    def _get_volume_group(self, array_mediator, volume_group_id):
+        try:
+            return array_mediator.get_volume_group(volume_group_id)
+        except array_errors.ObjectNotFoundError:
+            raise array_errors.ObjectNotFoundError(volume_group_id)
+
     @csi_method(error_response_type=csi_pb2.ModifyVolumeGroupResponse, lock_request_attribute="volume_group_id")
     def ModifyVolumeGroup(self, request, context):
-        raise NotImplementedError()
+        secrets = request.secrets
+        utils.validate_delete_volume_group_request(request)
+
+        try:
+            volume_group_id_info = utils.get_volume_group_id_info(request.volume_group_id)
+        except ObjectIdError as ex:
+            return handle_exception(ex, context, grpc.StatusCode.INVALID_ARGUMENT,
+                                    csi_pb2.ModifyVolumeGroupResponse)
+
+        array_type = volume_group_id_info.array_type
+        volume_group_id = volume_group_id_info.ids.uid
+        array_connection_info = utils.get_array_connection_info_from_secrets(secrets)
+
+        with get_agent(array_connection_info, array_type).get_mediator() as array_mediator:
+            logger.debug(array_mediator)
+
+            volume_group = self._get_volume_group(array_mediator, volume_group_id)
+
+            volumes_in_volume_group = volume_group.volumes
+            volume_ids_in_volume_group = [volume.id for volume in volumes_in_volume_group]
+            volume_ids_in_request = []
+            for volume_id in request.volume_ids:
+                volume_id_info = utils.get_volume_id_info(volume_id)
+                volume_ids_in_request.append(volume_id_info.ids.uid)
+
```

Comment:
Is it possible to add functions like 
volume_ids_in_request =build_volume_ids_in_request(request) 
and 
volume_ids_in_volume_group = build_volume_ids_in_group(volume_group)?

Reply:
for what reason? how is it different?

Reply:
I think it will be cleaner to move it into separate function. Why not?

Reply:
I agree with @ELENAGER

Reply:
I really don't understand, how is it gonna look?

Reply:
Something like this:
volume_ids_in_request = self._build_volume_ids_in_request(request)

Reply:
for where?
what in _build_volume_ids_in_request?
what line is it even for?
what is it building?

Reply:
I think we can create two functions:
```
def _get_volume_ids_in_request(self, request):
volume_ids_in_request = []
for volume_id in request.volume_ids:
    volume_id_info = utils.get_volume_id_info(volume_id)
    volume_ids_in_request.append(volume_id_info.ids.uid)
return volume_ids_in_request
```

```
def _get_volume_ids_in_volume_group(self, volume_group):
volumes_in_volume_group = volume_group.
return [volume.id for volume in volumes_in_volume_group]

```

Reply:
and in the `ModifyVolumeGroup` function, we will have two lines that each of them calls each function

Reply:
tnx, tnx